File: funtime/lib/floors-draw.js
/* ============================================================
   floors-draw.js — everything you can see in Hundred Floors

   Six kinds of platform, told apart by their PATTERN rather than by colour:
   a plain bar, a bar with 锯齿 sawteeth on top, arrows for the sliding ones, a
   coil for the spring, and a broken dashed bar for the one about to give way.

   Everything is drawn through screenY, because the world stays still and the
   camera is what moves.
   ============================================================ */

const COLOR_INK = '#111111';
const COLOR_PAPER = '#ffffff';
const COLOR_FAINT = '#d6d6d6';

/** clearCanvas — paint the whole canvas one flat colour. */
function clearCanvas(ctx, width, height, color) {
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, width, height);
}

/** drawFrame — the thin border round the shaft. */
function drawFrame(ctx, width, height) {
    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 2;
    ctx.strokeRect(1, 1, width - 2, height - 2);
}

/**
 * drawSawteeth — a row of 锯齿, the spikes that cost you blood.
 * INPUT: ctx, x, y, width, height, pointingUp. OUTPUT: nothing.
 * ALGORITHM: walk along in steps, drawing a triangle each time. The same
 *            function draws the platform spikes and the ceiling ones — the
 *            only difference is which way up they point.
 */
function drawSawteeth(ctx, x, y, width, height, pointingUp) {
    const step = 8;
    ctx.fillStyle = COLOR_INK;
    ctx.beginPath();
    for (let at = 0; at + step <= width; at = at + step) {
        if (pointingUp) {
            ctx.moveTo(x + at, y);
            ctx.lineTo(x + at + step / 2, y - height);
            ctx.lineTo(x + at + step, y);
        } else {
            ctx.moveTo(x + at, y);
            ctx.lineTo(x + at + step / 2, y + height);
            ctx.lineTo(x + at + step, y);
        }
        ctx.closePath();
    }
    ctx.fill();
}

/** drawCeiling — the spiked lid of the shaft, which you do not want to touch. */
function drawCeiling(ctx) {
    ctx.fillStyle = COLOR_INK;
    ctx.fillRect(0, 0, FIELD_WIDTH, CEILING_HEIGHT - 8);
    drawSawteeth(ctx, 0, CEILING_HEIGHT - 8, FIELD_WIDTH, 8, false);
}

/**
 * drawPlatform — one platform, drawn by its kind.
 * INPUT: ctx, state, platform. OUTPUT: nothing.
 */
function drawPlatform(ctx, state, platform) {
    const x = platform.x;
    const y = screenY(state, platform.y);
    const w = PLATFORM_WIDTH;
    const h = PLATFORM_HEIGHT;

    if (platform.kind === CRUMBLING) {
        /* a broken bar, and it fades as it gives way */
        ctx.strokeStyle = platform.crumbling > 0 ? COLOR_FAINT : COLOR_INK;
        ctx.lineWidth = 3;
        ctx.setLineDash([7, 5]);
        ctx.beginPath();
        ctx.moveTo(x, y + h / 2);
        ctx.lineTo(x + w, y + h / 2);
        ctx.stroke();
        ctx.setLineDash([]);
        return;
    }

    ctx.fillStyle = COLOR_INK;
    ctx.fillRect(x, y, w, h);

    if (platform.kind === SPIKED) {
        drawSawteeth(ctx, x, y, w, 9, true);

    } else if (platform.kind === SLIDE_LEFT || platform.kind === SLIDE_RIGHT) {
        const pointsRight = platform.kind === SLIDE_RIGHT;
        ctx.fillStyle = COLOR_PAPER;
        for (let i = 0; i < 3; i++) {
            const at = x + 12 + i * 18;
            ctx.beginPath();
            if (pointsRight) {
                ctx.moveTo(at, y + 2);
                ctx.lineTo(at + 7, y + h / 2);
                ctx.lineTo(at, y + h - 2);
            } else {
                ctx.moveTo(at + 7, y + 2);
                ctx.lineTo(at, y + h / 2);
                ctx.lineTo(at + 7, y + h - 2);
            }
            ctx.closePath();
            ctx.fill();
        }

    } else if (platform.kind === SPRING) {
        ctx.strokeStyle = COLOR_INK;
        ctx.lineWidth = 2;
        ctx.beginPath();
        ctx.moveTo(x + 10, y);
        for (let i = 0; i < 5; i++) {
            ctx.lineTo(x + 14 + i * 8, y - (i % 2 === 0 ? 8 : 0));
        }
        ctx.stroke();
    }
}

/**
 * drawPlayer — the little person.
 * ALGORITHM: a head, a body and two legs, with the legs apart while falling
 *            and together while standing, so you can see at a glance whether
 *            you are safely on something.
 */
function drawPlayer(ctx, state) {
    const player = state.player;
    const x = player.x;
    const y = screenY(state, player.y);
    const middle = x + PLAYER_WIDTH / 2;

    ctx.fillStyle = COLOR_INK;
    ctx.beginPath();
    ctx.arc(middle, y + 6, 6, 0, Math.PI * 2);
    ctx.fill();
    ctx.fillRect(middle - 5, y + 12, 10, 8);

    const apart = player.riding ? 2 : 5;
    ctx.fillRect(middle - apart - 2, y + 20, 3, 4);
    ctx.fillRect(middle + apart - 1, y + 20, 3, 4);

    ctx.fillStyle = COLOR_PAPER;
    ctx.fillRect(middle - 3, y + 4, 2, 2);
    ctx.fillRect(middle + 1, y + 4, 2, 2);
}

/**
 * drawHealth — the blood bar: one block per point left.
 * ALGORITHM: ten outlined boxes, filled in while you still have that much
 *            blood. No colour needed — full or empty says it all.
 */
function drawHealth(ctx, state) {
    const left = 10;
    const top = FIELD_HEIGHT - 22;
    const box = 11;

    ctx.font = 'bold 9px monospace';
    ctx.fillStyle = COLOR_INK;
    ctx.textAlign = 'left';
    ctx.fillText('BLOOD', left, top - 5);

    for (let i = 0; i < MAX_HEALTH; i++) {
        const x = left + i * (box + 2);
        ctx.strokeStyle = COLOR_INK;
        ctx.lineWidth = 1.5;
        ctx.strokeRect(x, top, box, box);
        if (i < state.health) {
            ctx.fillStyle = COLOR_INK;
            ctx.fillRect(x + 2, top + 2, box - 4, box - 4);
        }
    }
}

/** drawFloor — how far down you have got. */
function drawFloor(ctx, state) {
    ctx.fillStyle = COLOR_INK;
    ctx.font = 'bold 15px monospace';
    ctx.textAlign = 'right';
    ctx.fillText('FLOOR ' + state.floor, FIELD_WIDTH - 8, FIELD_HEIGHT - 10);
    ctx.textAlign = 'left';
}

/** drawMessage — big centred words across the shaft. */
function drawMessage(ctx, width, height, title, subtitle) {
    ctx.fillStyle = 'rgba(255, 255, 255, 0.93)';
    ctx.fillRect(0, height / 2 - 42, width, 84);
    ctx.fillStyle = COLOR_INK;
    ctx.textAlign = 'center';

    const scale = width / FIELD_WIDTH;
    const titleSize = Math.max(13, Math.round(26 * scale));
    const subtitleSize = Math.max(9, Math.round(12 * scale));

    ctx.font = 'bold ' + titleSize + 'px monospace';
    ctx.fillText(title, width / 2, height / 2 - 2);
    ctx.font = subtitleSize + 'px monospace';
    ctx.fillText(subtitle, width / 2, height / 2 + titleSize * 0.8);
    ctx.textAlign = 'left';
}

/** renderGame — draw one complete frame. */
function renderGame(ctx, state) {
    clearCanvas(ctx, FIELD_WIDTH, FIELD_HEIGHT, COLOR_PAPER);

    /* the walls of the shaft, so it is obvious you cannot leave sideways */
    ctx.strokeStyle = COLOR_FAINT;
    ctx.lineWidth = 4;
    ctx.beginPath();
    ctx.moveTo(2, 0);
    ctx.lineTo(2, FIELD_HEIGHT);
    ctx.moveTo(FIELD_WIDTH - 2, 0);
    ctx.lineTo(FIELD_WIDTH - 2, FIELD_HEIGHT);
    ctx.stroke();

    for (let i = 0; i < state.platforms.length; i++) {
        drawPlatform(ctx, state, state.platforms[i]);
    }

    drawPlayer(ctx, state);
    drawCeiling(ctx);
    drawHealth(ctx, state);
    drawFloor(ctx, state);
    drawFrame(ctx, FIELD_WIDTH, FIELD_HEIGHT);

    if (state.isOver) {
        const why = state.health <= 0 ? 'NO BLOOD LEFT' : 'YOU MISSED!';
        drawMessage(ctx, FIELD_WIDTH, FIELD_HEIGHT, why,
                    'floor ' + state.floor + ' — press R to try again');
    } else if (state.isPaused) {
        drawMessage(ctx, FIELD_WIDTH, FIELD_HEIGHT, 'PAUSED', 'Press P to carry on');
    }
}

File: funtime/pylib/floors_draw.py
"""floors_draw.py - everything you can see in Hundred Floors, in Python.

Six kinds of platform, told apart by their PATTERN rather than by colour: a
plain bar, a bar with 锯齿 sawteeth on top, arrows for the sliding ones, a coil
for the spring, and a broken dashed bar for the one about to give way.

Everything is drawn through screen_y, because the world stays still and the
camera is what moves.
"""

import math

from floors_rules import (FIELD_WIDTH, FIELD_HEIGHT, CEILING_HEIGHT,
                          PLAYER_WIDTH, PLATFORM_WIDTH, PLATFORM_HEIGHT,
                          MAX_HEALTH, PLAIN, SPIKED, SLIDE_LEFT, SLIDE_RIGHT,
                          SPRING, CRUMBLING, screen_y)

COLOR_INK = "#111111"
COLOR_PAPER = "#ffffff"
COLOR_FAINT = "#d6d6d6"


def clear_canvas(ctx, width, height, color):
    """Paint the whole canvas one flat colour."""
    ctx.fillStyle = color
    ctx.fillRect(0, 0, width, height)


def draw_frame(ctx, width, height):
    """The thin border round the shaft."""
    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 2
    ctx.strokeRect(1, 1, width - 2, height - 2)


def draw_sawteeth(ctx, x, y, width, height, pointing_up):
    """A row of 锯齿, the spikes that cost you blood.

    ALGORITHM: walk along in steps, drawing a triangle each time. The same
    function draws the platform spikes and the ceiling ones - the only
    difference is which way up they point.
    """
    step = 8
    ctx.fillStyle = COLOR_INK
    ctx.beginPath()
    at = 0
    while at + step <= width:
        if pointing_up:
            ctx.moveTo(x + at, y)
            ctx.lineTo(x + at + step / 2, y - height)
            ctx.lineTo(x + at + step, y)
        else:
            ctx.moveTo(x + at, y)
            ctx.lineTo(x + at + step / 2, y + height)
            ctx.lineTo(x + at + step, y)
        ctx.closePath()
        at += step
    ctx.fill()


def draw_ceiling(ctx):
    """The spiked lid of the shaft, which you do not want to touch."""
    ctx.fillStyle = COLOR_INK
    ctx.fillRect(0, 0, FIELD_WIDTH, CEILING_HEIGHT - 8)
    draw_sawteeth(ctx, 0, CEILING_HEIGHT - 8, FIELD_WIDTH, 8, False)


def draw_platform(ctx, state, platform):
    """One platform, drawn by its kind."""
    x = platform["x"]
    y = screen_y(state, platform["y"])
    w = PLATFORM_WIDTH
    h = PLATFORM_HEIGHT

    if platform["kind"] == CRUMBLING:
        # a broken bar, and it fades as it gives way
        ctx.strokeStyle = COLOR_FAINT if platform["crumbling"] > 0 else COLOR_INK
        ctx.lineWidth = 3
        ctx.setLineDash([7, 5])
        ctx.beginPath()
        ctx.moveTo(x, y + h / 2)
        ctx.lineTo(x + w, y + h / 2)
        ctx.stroke()
        ctx.setLineDash([])
        return

    ctx.fillStyle = COLOR_INK
    ctx.fillRect(x, y, w, h)

    if platform["kind"] == SPIKED:
        draw_sawteeth(ctx, x, y, w, 9, True)

    elif platform["kind"] in (SLIDE_LEFT, SLIDE_RIGHT):
        points_right = platform["kind"] == SLIDE_RIGHT
        ctx.fillStyle = COLOR_PAPER
        for i in range(3):
            at = x + 12 + i * 18
            ctx.beginPath()
            if points_right:
                ctx.moveTo(at, y + 2)
                ctx.lineTo(at + 7, y + h / 2)
                ctx.lineTo(at, y + h - 2)
            else:
                ctx.moveTo(at + 7, y + 2)
                ctx.lineTo(at, y + h / 2)
                ctx.lineTo(at + 7, y + h - 2)
            ctx.closePath()
            ctx.fill()

    elif platform["kind"] == SPRING:
        ctx.strokeStyle = COLOR_INK
        ctx.lineWidth = 2
        ctx.beginPath()
        ctx.moveTo(x + 10, y)
        for i in range(5):
            ctx.lineTo(x + 14 + i * 8, y - (8 if i % 2 == 0 else 0))
        ctx.stroke()


def draw_player(ctx, state):
    """The little person.

    ALGORITHM: a head, a body and two legs, with the legs apart while falling
    and together while standing, so you can see at a glance whether you are
    safely on something.
    """
    player = state["player"]
    x = player["x"]
    y = screen_y(state, player["y"])
    middle = x + PLAYER_WIDTH / 2

    ctx.fillStyle = COLOR_INK
    ctx.beginPath()
    ctx.arc(middle, y + 6, 6, 0, math.pi * 2)
    ctx.fill()
    ctx.fillRect(middle - 5, y + 12, 10, 8)

    apart = 2 if player["riding"] else 5
    ctx.fillRect(middle - apart - 2, y + 20, 3, 4)
    ctx.fillRect(middle + apart - 1, y + 20, 3, 4)

    ctx.fillStyle = COLOR_PAPER
    ctx.fillRect(middle - 3, y + 4, 2, 2)
    ctx.fillRect(middle + 1, y + 4, 2, 2)


def draw_health(ctx, state):
    """The blood bar: one block per point left."""
    left = 10
    top = FIELD_HEIGHT - 22
    box = 11

    ctx.font = "bold 9px monospace"
    ctx.fillStyle = COLOR_INK
    ctx.textAlign = "left"
    ctx.fillText("BLOOD", left, top - 5)

    for i in range(MAX_HEALTH):
        x = left + i * (box + 2)
        ctx.strokeStyle = COLOR_INK
        ctx.lineWidth = 1.5
        ctx.strokeRect(x, top, box, box)
        if i < state["health"]:
            ctx.fillStyle = COLOR_INK
            ctx.fillRect(x + 2, top + 2, box - 4, box - 4)


def draw_floor(ctx, state):
    """How far down you have got."""
    ctx.fillStyle = COLOR_INK
    ctx.font = "bold 15px monospace"
    ctx.textAlign = "right"
    ctx.fillText("FLOOR %d" % state["floor"], FIELD_WIDTH - 8, FIELD_HEIGHT - 10)
    ctx.textAlign = "left"


def draw_message(ctx, width, height, title, subtitle):
    """Big centred words across the shaft."""
    ctx.fillStyle = "rgba(255, 255, 255, 0.93)"
    ctx.fillRect(0, height / 2 - 42, width, 84)
    ctx.fillStyle = COLOR_INK
    ctx.textAlign = "center"

    scale = width / FIELD_WIDTH
    title_size = max(13, round(26 * scale))
    subtitle_size = max(9, round(12 * scale))

    ctx.font = "bold %dpx monospace" % title_size
    ctx.fillText(title, width / 2, height / 2 - 2)
    ctx.font = "%dpx monospace" % subtitle_size
    ctx.fillText(subtitle, width / 2, height / 2 + title_size * 0.8)
    ctx.textAlign = "left"


def render_game(ctx, state):
    """Draw one complete frame."""
    clear_canvas(ctx, FIELD_WIDTH, FIELD_HEIGHT, COLOR_PAPER)

    # the walls of the shaft, so it is obvious you cannot leave sideways
    ctx.strokeStyle = COLOR_FAINT
    ctx.lineWidth = 4
    ctx.beginPath()
    ctx.moveTo(2, 0)
    ctx.lineTo(2, FIELD_HEIGHT)
    ctx.moveTo(FIELD_WIDTH - 2, 0)
    ctx.lineTo(FIELD_WIDTH - 2, FIELD_HEIGHT)
    ctx.stroke()

    for platform in state["platforms"]:
        draw_platform(ctx, state, platform)

    draw_player(ctx, state)
    draw_ceiling(ctx)
    draw_health(ctx, state)
    draw_floor(ctx, state)
    draw_frame(ctx, FIELD_WIDTH, FIELD_HEIGHT)

    if state["is_over"]:
        why = "NO BLOOD LEFT" if state["health"] <= 0 else "YOU MISSED!"
        draw_message(ctx, FIELD_WIDTH, FIELD_HEIGHT, why,
                     "floor %d - press R to try again" % state["floor"])
    elif state["is_paused"]:
        draw_message(ctx, FIELD_WIDTH, FIELD_HEIGHT, "PAUSED", "Press P to carry on")
